Prose/ProseLib/Sources/ConversationFeature/Chat/Chat.swift

```diff
@@ -106,19 +214,137 @@ struct Chat: NSViewRepresentable {
+  func userContentController(
+    _: WKUserContentController,
+    didReceive message: WKScriptMessage
+  ) {
+    switch message.name {
+    case MessageMenuHandler.jsHandlerName:
+      MessageMenuHandler(viewStore: self.viewStore).handle(message)
+    case ToggleReactionHandler.jsHandlerName:
+      ToggleReactionHandler(viewStore: self.viewStore).handle(message)
+    case ShowReactionsHandler.jsHandlerName:
+      ShowReactionsHandler(viewStore: self.viewStore).handle(message)
+    case let name:
+      logger.info("Received message \(String(describing: name))")
+    }
+  }
+}
+
 struct ChatState: Equatable {
+  let loggedInUserJID: JID
+  let chatId: JID
   var isWebViewReady = false
   var messages = [Message]()
+  var selectedMessageId: Message.ID?
 }
 
 public enum ChatAction: Equatable {
   case webViewReady
+  case navigateUp, navigateDown
+  case didCreateMenu(MessageMenu, for: [Message.ID])
+  case messageMenuItemTapped(MessageMenu.Action)
+  case toggleReaction(Character, for: [Message.ID])
 }
 
-let chatReducer = Reducer<ChatState, ChatAction, Void> { state, action, _ in
+let chatReducer = Reducer<
+  ChatState,
+  ChatAction,
+  ConversationEnvironment
+> { state, action, environment in
   switch action {
   case .webViewReady:
     state.isWebViewReady = true
     return .none
+
+  case .navigateUp:
+    if let messageId = state.selectedMessageId,
+       let index = state.messages.lastIndex(where: { $0.id == messageId })
+    {
+      if index > 0 {
+        state.selectedMessageId = state.messages[index - 1].id
+      } else {
+        state.selectedMessageId = nil
+      }
+    } else {
+      state.selectedMessageId = state.messages.last?.id
+    }
+    return .none
+
+  case .navigateDown:
+    if let messageId = state.selectedMessageId,
+       let index = state.messages.lastIndex(where: { $0.id == messageId })
+    {
+      if index + 1 < state.messages.count {
+        state.selectedMessageId = state.messages[index + 1].id
+      } else {
+        state.selectedMessageId = nil
+      }
+    } else {
+      state.selectedMessageId = state.messages.first?.id
+    }
+    return .none
+
+  case let .didCreateMenu(menu, messageIds):
+    let loggedInUserJID: JID = state.loggedInUserJID
+
+    if let id = messageIds.first,
+       let message = state.messages.last(where: { $0.id == id }),
+       message.from == loggedInUserJID
+    {
+      return .fireAndForget {
+        // TODO: Uncomment once we support message edition
+//        menu.item(withTag: .edit)!.isEnabled = true
+        menu.item(withTag: .remove)!.isEnabled = true
+      }
+      .receive(on: environment.mainQueue)
+      .eraseToEffect()
+    }
+    return .none
+
+  case let .messageMenuItemTapped(action):
+    switch action {
+    case let .copyText(id):
+      logger.trace("Copying text of \(String(describing: id))…")
+      if let message = state.messages.last(where: { $0.id == id }) {
+        NSPasteboard.general.clearContents()
```

Comment:
That should be a new dependency since this is clearly a side-effect. That way we also can handle platform differences more easily.

```swift
public struct PasteboardClient {
  public var copyString: (String) -> Void

  public init(copyString: @escaping (String) -> Void) {
    self.copyString = copyString
  }
}

public extension PasteboardClient {
  static func live(pasteboard: UIPasteboard = .general) -> Self {
    .init(copyString: { pasteboard.string = $0 })
  }
}
```

Reply:
Ok, I will do that